### PR TITLE
Fix unstaking flows

### PR DESCRIPTION
**Pull request by jmoreira-valory**
This PR contains several improvements in terms of updating the staking/unstaking flow:

- Ensures that the agent bond on chain matches the selected staking contract.
- Collects metadata from staking programs on-chain.
- Fixes the staking/unstaking flow to match the discussions in the working document.
- Fixes computation of balance on `run_service.sh`
- Removes unused contract ABI JSONs from the folder `/contracts` and uses the files already available on the packages folder.
- Removes the `update_service.py` script and uses the built-in autonomy command.

### CODE REVIEW

File: scripts/utils.py [hunk 1]
```diff
@@ -69,7 +69,7 @@
 GAS_PARAMS = {
     "maxFeePerGas": 30_000_000_000,
     "maxPriorityFeePerGas": 3_000_000_000,
-    "gas": 500_000,
+    "gas": 5_000_000,
```
Comment (jmoreira-valory) [inline scripts/utils.py:72]: Verify

File: scripts/service_hash.py [hunk 2]
```diff
@@ -22,13 +22,19 @@
...
-SERVICE_ID_PATH = "../.trader_runner/service_id.txt"
-REGISTRY_JSON = "../contracts/ServiceRegistryL2.json"
+SCRIPT_PATH = Path(__file__).resolve().parent
+STORE_PATH = Path(SCRIPT_PATH, "..", ".trader_runner")
+DOTENV_PATH = Path(STORE_PATH, ".env")
+RPC_PATH = Path(STORE_PATH, "rpc.txt")
+SERVICE_ID_PATH = Path(STORE_PATH, "service_id.txt")
+PACKAGES_PATH = Path(SCRIPT_PATH, "..", "trader", "packages")
+
+REGISTRY_JSON = PACKAGES_PATH / "valory" / "contracts" / "service_registry" / "build" / "ServiceRegistryL2.json"
```
Comment (jmoreira-valory) [inline scripts/service_hash.py:37]: Remove redundant file from contracts (ServiceRegistryL2.json)

File: scripts/utils.py [hunk 3]
```diff
@@ -61,7 +61,6 @@
 DEFAULT_ON_CHAIN_INTERACT_SLEEP = 6.0
 
 
-ZERO_ADDRESS = "0x0000000000000000000000000000000000000000"
```
Comment (jmoreira-valory) [inline scripts/utils.py:64]: Already imported

File: terminate_on_chain_service.sh [hunk 4]
```diff
@@ -160,9 +160,8 @@ ask_password_if_needed
 cd trader
 validate_password
 
-if [ "${USE_STAKING}" = true ]; then
-    poetry run python "../scripts/staking.py" "$service_id" "$CUSTOM_SERVICE_REGISTRY_ADDRESS" "$CUSTOM_STAKING_ADDRESS" "../$operator_pkey_path" "$rpc" "$unstake";
-fi
+# Unstake if applicable
```
Comment (jmoreira-valory) [inline terminate_on_chain_service.sh:163]: The script already takes care if the service is staked or not.

File: contracts/AgentMech.json
Comment (jmoreira-valory) [inline contracts/AgentMech.json:1]: Not used anymore. It was used when the mech events were retrieved through the RPC. Now they are retrieved through the Subgraph.

File: contracts/ServiceRegistryL2.json
Comment (jmoreira-valory) [inline contracts/ServiceRegistryL2.json:1]: Already available on the packages folder of the service.

File: run_service.sh [hunk 5]
```diff
@@ -673,6 +673,12 @@ echo ""
 echo "This script will assist you in setting up and running the Trader service ($service_repo)."
 echo ""
 
+# Display information of the Git repository
+current_branch=$(git rev-parse --abbrev-ref HEAD)
+latest_commit_hash=$(git rev-parse HEAD)
+echo "Current branch: $current_branch"
+echo "Commit hash: $latest_commit_hash"
+
```
Comment (jmoreira-valory) [inline run_service.sh:681]: Help identify commit user is running when users request help.

File: run_service.sh [hunk 6]
```diff
@@ -880,15 +886,26 @@ remote_service_hash=$(poetry run python "../scripts/service_hash.py")
...
+# On-chain agent bond for the expected agent ID ($AGENT_ID)
+on_chain_agent_bond=$(poetry run python "../scripts/get_agent_bond.py" "$CUSTOM_SERVICE_REGISTRY_TOKEN_UTILITY_ADDRESS" "$service_id" "$AGENT_ID" "$rpc")
+
+if [ "${USE_STAKING}" = true ]; then
+    cost_of_bonding=$MIN_STAKING_BOND_OLAS
+else
+    cost_of_bonding=$MIN_STAKING_BOND_XDAI
+fi
+
+if [ "$local_service_hash" != "$remote_service_hash" ] || [ "$on_chain_agent_id" != "$AGENT_ID" ] || [ "$on_chain_agent_bond" != "$cost_of_bonding" ]; then
```
Comment (jmoreira-valory) [inline run_service.sh:898]: If hash, agent_id or bond don't match the expected values, it requires an on-chain service update.

File: run_service.sh [hunk 7]
```diff
@@ -971,25 +988,26 @@ if [ "$local_service_hash" != "$remote_service_hash" ] || [ "$on_chain_agent_id"
...
-                  --use-custom-chain \
-                  service packages/valory/services/trader/ \
-                  --key \"../$operator_pkey_path\" $password_argument \
-                  --nft $nft \
-                  -a $AGENT_ID \
-                  -n $n_agents \
-                  -c $cost_of_bonding \
-                  --threshold $n_agents \
-                  --update \"$service_id\""
+              cmd+=" -c $cost_of_bonding"
```
Comment (jmoreira-valory) [inline run_service.sh:1008]: Use the already available autonomy command (`--token` option was not available and a custom script was created)

File: scripts/choose_staking.py [hunk 8]
```diff
@@ -25,85 +25,52 @@
...
-        "description": "The Quickstart Beta - Expert staking contract offers 20 slots for operators running Olas Predict agents with the quickstart. It is designed for professional agent operators, requiring 1000 OLAS for staking. The rewards are proportional to the Quickstart Beta - Hobbyist.",
-        # https://github.com/valory-xyz/autonolas-staking-programmes/blob/main/scripts/deployment/globals_gnosis_mainnet_qs_beta_expert.json
-        "deployment": {
-            "stakingTokenAddress": "0xEa00be6690a871827fAfD705440D20dd75e67AB1",
-            "stakingTokenInstanceAddress": "0x5344B7DD311e5d3DdDd46A4f71481bD7b05AAA3e"
-        }
-    },
+    NO_STAKING_PROGRAM_ID: ZERO_ADDRESS,
+    "quickstart_beta_hobbyist": "0x389B46c259631Acd6a69Bde8B6cEe218230bAE8C",
+    "quickstart_beta_expert": "0x5344B7DD311e5d3DdDd46A4f71481bD7b05AAA3e",
```
Comment (jmoreira-valory) [inline scripts/choose_staking.py:67]: Simplified this data structure - we only need the contract address. All remaining information can be fetch on-chain.

**Comment by jmoreira-valory** [comment]
Resolves #151

**Review by Adamantios — APPROVED** [review]

File: scripts/utils.py (hunk 1 above)
Comment (jmoreira-valory) [reply scripts/utils.py]: @0xArdi `"gas": 500_000` has caused issues sometimes on Tenderly when sending the unstake transaction. I suspect it usually runs out of gas due to being executed when several checkpoints have not been called (thus it needs to do extra work). This seems not happening on mainnet, where the checkpoint is periodically called. What do you suggest to do here?

Comment (0xArdi) [reply scripts/utils.py]: Ideally we do this using automatic gas estimation.

Comment (jmoreira-valory) [reply scripts/utils.py]: Yes, but in this script we are not doing it. So what should we do? keep it asi it is 500.000 ? increase it?

**Review by 0xArdi — APPROVED** [review]

**Pull request merged**